### PR TITLE
Add allowed_run_types to whitelist specific dag run types

### LINKED ISSUE #61710: Add the option to limit specific dag run types
### Body

Some DAGs have strict logic of using intervals for processing (for example `{{ prev_data_interval_start_success }}` to `{{ data_interval_end }}` ). For such DAGs invoking manual runs will cause a problem as manual runs behave differently with intervals and scheduled runs do not reference them.
This mostly becomes an issue when on-call who is not the DAG authors try to fix an issue or recover from a problem and isn't fully aware of the recover procedure for this specific DAG (need to clear a failure rather than triggering a new manual run).

To avoid this problem I suggest to give Dag authors the ability to decide which dag run types are acceptable.
Something like:

```
from airflow.utils.types import DagRunType
my_dag = DAG(
    ...,
    deny_dag_run_types = [DagRunType.MANUAL]
)
```

Where the default of `deny_dag_run_types` is to accept all run types. The interface/parameter name is just a suggestion.

The expected behavior is for the UI/API to show a message that manual runs are not supported for this dag.

After discussing with @guan404ming we'll need to break it down to the following:
- Model: Add a deny_dag_run_types field to the dag model + DB migration.
- API: Add validation in the trigger dag run endpoint to reject denied run types.
- UI: Disable/hide the trigger button when MANUAL is denied, maybe with a tooltip explaining why.

There may be edge cases that needs to raise broken Dag message. For example:

```
from airflow.utils.types import DagRunType
my_dag = DAG(
    ...,
    schedule=[example_asset]
    deny_dag_run_types = [DagRunType.ASSET_TRIGGERED]
)
```
as it make no sense to block asset triggering runs when the scheduling is per asset.

### Committer

- [x] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by guan404ming**
<!--
Thank you for contributing!

Please provide above a brief description of the changes made in this pull request.
Write a good git commit message following this guide: http://chris.beams.io/posts/git-commit/

Please make sure that your code changes are covered with tests.
And in case of new features or big changes remember to adjust the documentation.

Feel free to ping (in general) for the review if you do not see reaction for a few days
(72 Hours is the minimum reaction time you can expect from volunteers) - we sometimes miss notifications.

In case of an existing issue, reference it using one of the following:

* closes: #ISSUE
* related: #ISSUE
-->

## Related Issue

Closes #61710

## Why                                                                                                                                                       
                                                                                                                                                          
Allow dag authors to restrict which run types (manual, scheduled, backfill, asset-triggered) are permitted on a per-DAG basis.                            
                                                                                                                                                          
## How

- Add allowed_run_types field to definition, DagModel, serialization, and a DB migration
- Enforce the allow list in scheduler, trigger, backfill, and asset materialization API routes
- Disable the trigger button in the UI when manual runs are not allowed

<img width="2914" height="1544" alt="image" src="https://github.com/user-attachments/assets/d31cd427-4303-4e14-8f61-beaa1ba803bd" />


---

##### Was generative AI tooling used to co-author this PR?

<!--
If generative AI tooling has been used in the process of authoring this PR, please
change below checkbox to `[X]` followed by the name of the tool, uncomment the "Generated-by".
-->

- [x] Yes (please specify the tool below)
Claude Code with Opus 4.6

<!--
Generated-by: [Tool Name] following [the guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)
-->

---

* Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information. Note: commit author/co-author name and email in commits become permanently public when merged.
* For fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
* When adding dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
* For significant user-facing changes create newsfragment: `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

### CODE REVIEW

**Comment by eladkal** [comment]
Great work!
I'm having an issue with trying to run:


```
dag = DAG(
    ...
    schedule="@daily",
    deny_dag_run_types=DagRunType.MANUAL
)
```

I get

```
Traceback (most recent call last):
  File "/usr/python/lib/python3.11/enum.py", line 714, in __call__
    return cls.__new__(cls, value)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/python/lib/python3.11/enum.py", line 1137, in __new__
    raise ve_exc
ValueError: 'm' is not a valid DagRunType
```

It's solved if I change `deny_dag_run_types=[DagRunType.MANUAL]`.
I think we should allow also a single value... It makes things easier and shouldn't be too hard to fix

File: task-sdk/src/airflow/sdk/definitions/dag.py
Comment (eladkal) [inline task-sdk/src/airflow/sdk/definitions/dag.py:1]: Missing

`:param deny_dag_run_types` in the DAG class docstring

**Review by eladkal — COMMENTED** [review]
I think we are missing tests for cases like:

```
from airflow.utils.types import DagRunType
my_dag = DAG(
    ...,
    schedule=[example_asset]
    deny_dag_run_types = [DagRunType.ASSET_TRIGGERED]
)
```

```
from airflow.utils.types import DagRunType
my_dag = DAG(
    ...,
    schedule="@daily"
    deny_dag_run_types = [DagRunType.SCHEDULED]
)
```

```
from airflow.utils.types import DagRunType
my_dag = DAG(
    ...,
    schedule=None
    deny_dag_run_types = [DagRunType.MANUAL]
)
```

Test should verify broken dag message is raised for all these 3 cases

**Comment by guan404ming** [comment]
> Test should verify broken dag message is raised for all these 3 cases

Thanks, just updated to make the tests covering more cased.

> It's solved if I change deny_dag_run_types=[DagRunType.MANUAL]. I think we should allow also a single value... It makes things easier and shouldn't be too hard to fix

Sure, that is a nice suggestion. I've updated to support that and also fix some type issue there.

File: task-sdk/src/airflow/sdk/definitions/dag.py
Comment (guan404ming) [reply task-sdk/src/airflow/sdk/definitions/dag.py:1]: Thanks for the great catch. Just updated!

**Review by shahar1 — COMMENTED** [review]
Looks like a valuable feature!
1. For enhanced security, I'd make it an allowlist rather than disallowlist. If we theoretically add a new Dag type in the future, this ensures that it does not get inside by mistake.
2. I'd name it something like `(dis)allowed_run_types ` rather than `deny/approve_run_types`, to reflect that it's an attribute of the Dag - using the verb `deny`/`allow` is a bit confusing.
3. It should be clarified and documented how it behaves when a Dag type, or the entire parameter, is removed (e.g., does it respect the `catchup` parameter?).

**Comment by eladkal** [comment]
> It should be clarified and documented how it behaves when a Dag type, or the entire parameter, is removed (e.g., does it respect the catchup parameter?).

I think the logic is pretty simple. This parameter prevents creation of runs. There are no exceptions thus if you prevent schedule runs all the settings that specify which scheduled runs should be created are ignored.

**Comment by shahar1** [comment]
> > It should be clarified and documented how it behaves when a Dag type, or the entire parameter, is removed (e.g., does it respect the catchup parameter?).
> 
> I think the logic is pretty simple. This parameter prevents creation of runs. There are no exceptions thus if you prevent schedule runs all the settings that specify which scheduled runs should be created are ignored.

I'll try to clarify - 
Let's say that using this parameter you allow only asset runs (`allowed_run_types = ['assets']`).
At some future point you decide that you want to re-add scheduled runs (`allowed_run_types =['assets', 'scheduled']`).

1. Do we respect the `catchup` parameter in this case to complete past runs? (see followup)
2. If we do, are **all** past scheduled runs queued, or only those between setting the initial `allowed_run_types` and updating it? You might had set `catchup=False` before adding `allowed_run_types`, and if you now change both `catchup=True` and allow scheduled runs, there will be a lot of potentially unnecessary runs for the users.

We could keep it simple for now and make the required assumptions, but it should be documented IMO.

**Comment by eladkal** [comment]
> I'll try to clarify - Let's say that using this parameter you allow only asset runs (`allowed_run_types = ['assets']`). At some future point you decide that you want to re-add scheduled runs (`allowed_run_types =['assets', 'scheduled']`).
> 
> 1. Do we respect the `catchup` parameter in this case to complete past runs? (see followup)
> 2. If we do, are **all** past scheduled runs queued, or only those between setting the initial `allowed_run_types` and updating it? You might had set `catchup=False` before adding `allowed_run_types`, and if you now change both `catchup=True` and allow scheduled runs, there will be a lot of potentially unnecessary runs for the users.
> 
> We could keep it simple for now and make the required assumptions, but it should be documented IMO.

Catachup is not the same as backfill.
Catchup works only for dates later than the latest existed dagrun. It does not consider dates before the latest dagrun. In your example changing to catchup=true will do nothing regardless of the new flag.

**Review by eladkal — APPROVED** [review]
LGTM
I agree with @shahar1 on points 1-2 in https://github.com/apache/airflow/pull/61833#pullrequestreview-3807621037 better to apply them

**Comment by guan404ming** [comment]
Thanks for the suggestions and discussion
                                                                                                                                                        
1. Agree that an allowlist (whitelist) is more reasonable from a security perspective. It may require a bit more effort from users, but I think the safety tradeoff is worth it.                            
2. Good point on the naming. Will rename to `allowed_run_types`.

File: airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py [hunk 1]
```diff
@@ -1541,8 +1541,20 @@ def _dags_for_trigger_tests(self, session=None):
         )
         import_errors_dag.has_import_errors = True
 
+        deny_manual_dag = DagModel(
+            dag_id="deny_manual",
+            bundle_name="testing",
+            fileloc="/tmp/dag_del_3.py",
```
Comment (shahar1) [inline airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py]: nit:
```suggestion
        allowed_scheduled_dag = DagModel(
            dag_id="allowed_scheduled",
            bundle_name="testing",
            fileloc="/tmp/dag_del_3.py",
```

File: airflow-core/src/airflow/jobs/scheduler_job_runner.py [hunk 2]
```diff
@@ -1810,6 +1810,18 @@ def _create_dagruns_for_partitioned_asset_dags(self, session: Session) -> set[st
...
+            if (
+                dag_model
+                and dag_model.allowed_run_types is not None
+                and DagRunType.ASSET_TRIGGERED.value not in dag_model.allowed_run_types
+            ):
+                self.log.warning(
+                    "Dag does not allow asset-triggered runs; skipping",
+                    dag_id=apdr.target_dag_id,
+                )
+                continue
```
Comment (jason810496) [inline airflow-core/src/airflow/jobs/scheduler_job_runner.py]: It seems we could modularize the check and the warning.

**Review by shahar1 — APPROVED** [review]
LGTM, don't forget to update PR's title + commit message before merging :)

File: airflow-core/src/airflow/models/backfill.py [hunk 3]
```diff
@@ -100,6 +100,14 @@ class InvalidBackfillDate(AirflowException):
     """
 
 
+class DeniedDagRunType(AirflowException):
```
Comment (jason810496) [inline airflow-core/src/airflow/models/backfill.py]: Maybe `models.dagrun` or `models.dag` module is a better place for `DeniedDagRunType`?

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py [hunk 4]
```diff
@@ -402,6 +403,13 @@ def materialize_asset(
             f"More than one DAG materializes asset with ID: {asset_id}",
         )
 
+    dm = session.scalar(select(DagModel).where(DagModel.dag_id == dag_id).limit(1))
+    if dm and dm.allowed_run_types is not None and DagRunType.MANUAL.value not in dm.allowed_run_types:
```
Comment (jason810496) [inline airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: Not sure whether we should treat ‎`materialize_asset` as ‎`DagRunType.ASSET_TRIGGERED` or ‎`DagRunType.MANUAL` 👀

File: airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py [hunk 5]
```diff
@@ -112,6 +112,15 @@ def trigger_dag_run(
...
 
+    if dm.allowed_run_types is not None and DagRunType.MANUAL.value not in dm.allowed_run_types:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
+            detail={
+                "reason": "denied_run_type",
+                "message": f"Dag with dag_id '{dag_id}' does not allow manual runs",
+            },
+        )
+
```
Comment (jason810496) [inline airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py]: There’s an edge case where the ‎`TriggerDagRunOperator` task will also call the ‎`trigger_dag_run` route, but we can also choose not to set ‎`manual` in ‎`allowed_run_types` in the same Dag.

File: airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py (hunk 1 above)
Comment (guan404ming) [reply airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py]: Thanks just updated!

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py [hunk 6]
```diff
@@ -402,6 +403,13 @@ def materialize_asset(
             f"More than one DAG materializes asset with ID: {asset_id}",
         )
 
+    dm = session.scalar(select(DagModel).where(DagModel.dag_id == dag_id).limit(1))
```
Comment (ashb) [inline airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: I'm not 100% sure this is right for this route?

Ptal @Lee-W

Comment (ashb) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: I almost wonder if those should be a different type to "manual" here

File: airflow-core/src/airflow/dag_processing/collection.py [hunk 7]
```diff
@@ -600,6 +601,14 @@ def update_dags(
             dm.timetable_description = dag.timetable.description
             dm.fail_fast = dag.fail_fast if dag.fail_fast is not None else False
 
+            allowed_types = dag.allowed_run_types
+            if allowed_types:
+                dm.allowed_run_types = sorted(
+                    v.value if isinstance(v, enum.Enum) else v for v in allowed_types
+                )
```
Comment (ashb) [inline airflow-core/src/airflow/dag_processing/collection.py]: This looks a bit odd. What stop us doing

```suggestion
                dm.allowed_run_types = sorted(allowed_types)
```

File: airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+down_revision = "f8c9d7e6b5a4"
+branch_labels = None
+depends_on = None
+airflow_version = "3.2.0"
+
+
+def upgrade():
+    """Add allowed_run_types column to dag table."""
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("allowed_run_types", sa.JSON(), nullable=True))
```
Comment (ashb) [inline airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py:44]: Json is overkill/allows more than we want here - maybe for postgres this would be better as the native ARRAY type?

File: airflow-core/src/airflow/jobs/scheduler_job_runner.py [hunk 9]
```diff
@@ -2060,6 +2082,16 @@ def _create_dag_runs_asset_triggered(
                 )
                 continue
 
+            if (
+                dag_model.allowed_run_types is not None
+                and DagRunType.ASSET_TRIGGERED.value not in dag_model.allowed_run_types
+            ):
+                self.log.warning(
+                    "Dag does not allow asset-triggered runs; skipping",
```
Comment (ashb) [inline airflow-core/src/airflow/jobs/scheduler_job_runner.py]: This shouldn't be a skip here - i think this should be a parse time error when you try to reference a dag from an asset without allowing the type

File: airflow-core/src/airflow/models/backfill.py (hunk 3 above)
Comment (ashb) [inline airflow-core/src/airflow/models/backfill.py]: This is nothing to do with backfills, so this should be elsewhere

File: airflow-core/src/airflow/models/backfill.py [hunk 10]
```diff
@@ -502,6 +510,14 @@ def _create_backfill(
         if not serdag:
             raise DagNotFound(f"Could not find dag {dag_id}")
 
+        dag_model = session.scalar(select(DagModel).where(DagModel.dag_id == dag_id).limit(1))
```
Comment (ashb) [inline airflow-core/src/airflow/models/backfill.py:505]: Is there any way we could combine this and the next select on to one?

File: airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py (hunk 8 above)
Comment (guan404ming) [reply airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py:44]: The codebase currently uses `sa.JSON()` for all list-type columns, e.g. `assignees` and `chosen_options` in the HITL migration. I found there is no native `ARRAY` usage anywhere in models or migrations. Kept it as JSON here to stay consistent, but I am also happy to open a follow-up issue to explore switching these to native ARRAY where supported if that's preferred.

File: task-sdk/src/airflow/sdk/definitions/dag.py [hunk 11]
```diff
@@ -590,6 +610,29 @@ def _validate_tags(self, _, tags: Collection[str]):
         if tags and any(len(tag) > TAG_MAX_LEN for tag in tags):
             raise ValueError(f"tag cannot be longer than {TAG_MAX_LEN} characters")
 
+    @allowed_run_types.validator
+    def _validate_allowed_run_types(self, _, allowed_run_types):
+        if not allowed_run_types:
+            return
+        from airflow.utils.types import DagRunType
```
Comment (jason810496) [inline task-sdk/src/airflow/sdk/definitions/dag.py]: Maybe import from `airflow.sdk.api.datamodels._generated` instead of airflow-core?

**Review by jason810496 — COMMENTED** [review]
Nice! LGTM overall and it's really a nice improvement for users.

Updated:
Sorry, there're some outdated comments same as Ash already pointed out, I just resolved them.

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py (hunk 4 above)
Comment (jason810496) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: Same as Ash pointed out, resolved.

File: airflow-core/src/airflow/models/backfill.py (hunk 3 above)
Comment (jason810496) [reply airflow-core/src/airflow/models/backfill.py]: Same as Ash pointed out, resolved.

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py (hunk 6 above)
Comment (guan404ming) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: > I'm not 100% sure this is right for this route?

I updated with moving the check after `get_latest_version_of_dag` and use `dag.allowed_run_types` directly from the `SerializedDAG`, removing the extra `DagModel` query.

Comment (guan404ming) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: > I almost wonder if those should be a different type to "manual" here

Agree this is worth considering. For now it creates a DagRunType.MANUAL run so the check is consistent, but a dedicated materialization run type could make sense as a follow-up.

**Comment by guan404ming** [comment]
Thanks for all the comments. I just updated with @ashb review comments and will handle @jason810496's tomorrow!

File: airflow-core/src/airflow/models/backfill.py (hunk 10 above)
Comment (guan404ming) [reply airflow-core/src/airflow/models/backfill.py:505]: Yes, I found that we could reused the already-fetched `dag_model` to check `dag_model.timetable_summary` == "None" instead of a separate `select(func.count())` query. Combined into one if `dag_model:` block with both checks.

File: airflow-core/src/airflow/models/backfill.py (hunk 3 above)
Comment (guan404ming) [reply airflow-core/src/airflow/models/backfill.py]: Thanks.

File: airflow-core/src/airflow/jobs/scheduler_job_runner.py (hunk 9 above)
Comment (guan404ming) [reply airflow-core/src/airflow/jobs/scheduler_job_runner.py]: Make sense to me, just moved to the parse time check.

File: task-sdk/src/airflow/sdk/definitions/dag.py (hunk 11 above)
Comment (guan404ming) [reply task-sdk/src/airflow/sdk/definitions/dag.py]: Good suggestion. However the existing code in this file already imports `DagRunType` from `airflow.utils.types` (line 1222), so this is consistent with the current pattern. Also, the two `DagRunType` enums have different member names (`BACKFILL_JOB` vs `BACKFILL`), so switching would need some care. Maybe better addressed as a separate cleanup across the file. What do you think?

File: airflow-core/src/airflow/jobs/scheduler_job_runner.py (hunk 2 above)
Comment (guan404ming) [reply airflow-core/src/airflow/jobs/scheduler_job_runner.py]: This check has been removed. Parse-time validation `in _validate_allowed_run_types` already raises a ValueError if an asset-scheduled dag doesn't include `ASSET_TRIGGERED` in `allowed_run_types`, so the runtime check was redundant.

File: airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py (hunk 5 above)
Comment (guan404ming) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py]: Good catch. Currently `TriggerDagRunOperator` creates `MANUAL` runs, so the check is consistent with the current behavior. But you're right that operator-triggered runs are semantically different from user-initiated manual runs. A dedicated run type for operator-triggered runs could address this, but that's a larger change. I just added a TODO comment for now and I think I could follow up one to update more dedicated run type here.

File: task-sdk/src/airflow/sdk/definitions/dag.py (hunk 11 above)
Comment (amoghrajesh) [reply task-sdk/src/airflow/sdk/definitions/dag.py]: We should avoid using imports from `airflow-core` in task sdk going forward. It is best to use `airflow.sdk.api.datamodels._generated` as @jason810496 mentioned.

> However the existing code in this file already imports DagRunType from airflow.utils.types 

Regarding this, we plan to get rid of it in these ticket: https://github.com/apache/airflow/issues/54713

Switching should not be too hard, just do it here, the other refs can be handled in the related ticket

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py [hunk 12]
```diff
@@ -459,6 +459,12 @@ def trigger_dag_run(
             f"DAG with dag_id: '{dag_id}' has import errors and cannot be triggered",
         )
 
+    if dm.allowed_run_types is not None and DagRunType.MANUAL.value not in dm.allowed_run_types:
```
Comment (roykoand) [inline airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py]:
```suggestion
    if dm.allowed_run_types is not None and DagRunType.MANUAL not in dm.allowed_run_types:
```

Wouldn't it work without a `value`? It seems like accessing enum value throught the pr is not consistent

Comment (guan404ming) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/dag_run.py]: Make sense, thanks. Just updated.

File: task-sdk/src/airflow/sdk/definitions/dag.py (hunk 11 above)
Comment (guan404ming) [reply task-sdk/src/airflow/sdk/definitions/dag.py]: Just updated, thanks!

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py (hunk 6 above)
Comment (eladkal) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: Shall we open a followup task or leave a todo with instructions in the code? Otherwise we wil lose track of it

Comment (guan404ming) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: Just opened. Feel free to drop any options there #62125. Thanks!

**Comment by eladkal** [comment]
@guan404ming did you address all comments? are we ready to merge?

**Comment by guan404ming** [comment]
Yes, I think I have resolved all the comments and we could refine it if there is any issue after merging.

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/backfills.py [hunk 13]
```diff
@@ -252,6 +252,11 @@ def create_backfill(
             status_code=status.HTTP_404_NOT_FOUND,
             detail=f"Could not find dag {backfill_request.dag_id}",
         )
+    except DagRunTypeNotAllowed as e:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
```
Comment (jason810496) [inline airflow-core/src/airflow/api_fastapi/core_api/routes/public/backfills.py:259]: We need to update the `create_openapi_http_exception_doc` for the route definition.

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py [hunk 14]
```diff
@@ -403,6 +403,13 @@ def materialize_asset(
         )
 
     dag = get_latest_version_of_dag(dag_bag, dag_id, session)
+
+    if dag.allowed_run_types is not None and DagRunType.MANUAL not in dag.allowed_run_types:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
```
Comment (jason810496) [inline airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py:411]: Same here.

File: airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py [hunk 15]
```diff
@@ -112,6 +112,17 @@ def trigger_dag_run(
             },
         )
 
+    # TODO: TriggerDagRunOperator also calls this route but creates MANUAL runs.
+    #  Consider a dedicated run type for operator-triggered runs.
+    if dm.allowed_run_types is not None and DagRunType.MANUAL not in dm.allowed_run_types:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
```
Comment (jason810496) [inline airflow-core/src/airflow/api_fastapi/execution_api/routes/dag_runs.py:119]: We need to update the description of `HTTP_400_BAD_REQUEST` about `denied_run_type`.

File: airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py (hunk 8 above)
Comment (jason810496) [reply airflow-core/src/airflow/migrations/versions/0104_3_2_0_add_allowed_run_types_to_dag.py:44]: I had the same thought as  Ash during last review, maybe we could track this in a new issue about the ARRAY <-> JSON trade-off for list-type columns.

File: airflow-core/src/airflow/ui/public/i18n/locales/en/components.json [hunk 16]
```diff
@@ -130,6 +130,7 @@
     "intervalStart": "Start",
     "loading": "Loading Dag information...",
     "loadingFailed": "Failed to load Dag information. Please try again.",
+    "manualRunDenied": "Manual runs are not allowed for this DAG",
```
Comment (jason810496) [inline airflow-core/src/airflow/ui/public/i18n/locales/en/components.json]:
```suggestion
    "manualRunDenied": "Manual runs are not allowed for this Dag",
```

File: task-sdk/src/airflow/sdk/definitions/dag.py [hunk 17]
```diff
@@ -590,6 +610,29 @@ def _validate_tags(self, _, tags: Collection[str]):
...
+                )
+        elif self.timetable.can_be_scheduled:
+            if DagRunType.SCHEDULED not in allowed_run_types:
+                raise ValueError(
+                    "allowed_run_types must include SCHEDULED when the DAG has a schedule defined"
+                )
+        else:
+            if DagRunType.MANUAL not in allowed_run_types:
+                raise ValueError(
+                    "allowed_run_types must include MANUAL when the DAG "
```
Comment (jason810496) [inline task-sdk/src/airflow/sdk/definitions/dag.py]:
```suggestion
                    "allowed_run_types must include ASSET_TRIGGERED when the Dag is scheduled by assets"
                )
        elif self.timetable.can_be_scheduled:
            if DagRunType.SCHEDULED not in allowed_run_types:
                raise ValueError(
                    "allowed_run_types must include SCHEDULED when the Dag has a schedule defined"
                )
        else:
            if DagRunType.MANUAL not in allowed_run_types:
                raise ValueError(
                    "allowed_run_types must include MANUAL when the Dag "
```

**Review by jason810496 — COMMENTED** [review]
Nice! Thanks for the rapid update!

**Comment by guan404ming** [comment]
Hi @jason810496, thanks for your great catch. I just updated with your comments.

**Review by jason810496 — APPROVED** [review]
Thanks @guan404ming!

**Comment by guan404ming** [comment]
Thanks!

File: airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py (hunk 6 above)
Comment (Lee-W) [reply airflow-core/src/airflow/api_fastapi/core_api/routes/public/assets.py]: Late to the discussion. It's ok to be manual, but it might be better to have a separate type. now reviewing the PR
